README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# wmts

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/mdsumner/wmts.svg?branch=master)](https://travis-ci.org/mdsumner/wmts)
<!-- badges: end -->

The goal of wmts is to obtain imagery from online image servers. You can
provide *any URL* with a `GetCapabilities.xml` and quickly get an image
for any region at any resolution.

*Any URL*? Only Pseudo Mercator services tested so far, see examples
below for success stories. Please [file an
issue](https://github.com/mdsumner/wmts/issues) or ping me [on
twitter](https://twitter.com/mdsumner) if you have
wants/needs/suggestions/bug-fixes/high-fives.

## The wmts() function

Use `loc` to specify a location (`cbind(lon, lat)`) and `buffer` to give
a radius around that point in metres. Alternatively, any spatial object
can be used as `loc` and its extent (in the Mercator tile server
projection) will be used.

There is an argument `max_tiles` which is 9 by default (3x3 256x256
tiles) and ensures a reasonably small sized result. Increase max\_tiles
to obtain greater resolution (squares make sense, i.e. 16, 25, 36, 49, …
but the actual shape depends on `loc` and `buffer`).

The `zoom` argument can be used but please take care, the function will
report the zoom in use and it’s advisable not to increase this much (use
max\_tiles to get a known maximum amount of data).

See [GDAL WMTS documentation](https://gdal.org/drivers/raster/wmts.html)
for *technical details* on this GDAL-based approach.

## Beware

  - will create temporary files (gdalwmscache, and raster temp)
  - no checks for in-bounds for tiles
  - no safety checks if you manually set a large zoom
  - no idea what systems/OS this works on yet (tested on Ubuntu)

## TODO

  - \[x\] need to deal with various inputs
  - \[x\] fix URL forms (http/s vs WMTS:)
  - \[ \] add providers

## Example

``` r
## remotes::install_github("mdsumner/wmts")
centre <- c(-80.888, 32.332)  ## lonlat
radius <- 4000                ## metres
u <- "WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer=USGSTopo,tilematrixset=default028mm"
library(wmts)

## note that input can be centre (longlat pt), buffer (km) or a spatial object (ignore buffer)
## output is a raster in World Mercator
x <- wmts(u, centre, buffer = radius) ## zoom determined interactively, don't use a hardcoded zoom
#> zoom: 13
raster::plotRGB(x, interpolate = TRUE)
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r
#f <- system.file("gpkg/nc.gpkg", package = "sf", mustWork = TRUE)
#sf <- sf::read_sf(f)
#x <- wmts(u, sf)
tab <- tibble::as_tibble(raster::as.data.frame(x, xy = TRUE))
names(tab) <- c("x", "y", "red", "green", "blue")
##tab <- dplyr::filter(tab, !is.na(red))  ##... when we have missing values, we should drop them or rgb() will error
tab$hex <- rgb(tab$red, tab$green, tab$blue, maxColorValue = 255)
library(ggplot2)
ggplot(tab, aes(x, y, fill = hex)) +
  geom_raster(interpolate = TRUE) +
  coord_equal() +
  scale_fill_identity()
```

<img src="man/figures/README-example-2.png" width="100%" />

## Tasmania\!

Get a Tassie example.

``` r
ext <- raster::extent(146, 148, -43, -40.5)
u <- "WMTS:https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/Topographic/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer=Basemaps_Topographic,tilematrixset=default028mm"

img <- wmts(u, loc = ext)
#> zoom: 8
raster::plotRGB(img, interpolate = TRUE)
```

<img src="man/figures/README-tassie-1.png" width="100%" />

``` r

ortho <- wmts("https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/Orthophoto/MapServer/WMTS/1.0.0/WMTSCapabilities.xml", raster::extent(147.125, 147.528, -43.054, -42.717), max_tiles = 16)
#> zoom: 11
raster::plotRGB(ortho, interpolate = TRUE)
```

<img src="man/figures/README-tassie-2.png" width="100%" />

``` r


## note that we can use an spatial object for the extent, saving much pain
url_shade <- "https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/HillshadeGrey/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"
greyshade <- wmts(url_shade, ortho)
#> zoom: 10
raster::plotRGB(greyshade, interpolate = TRUE)
```

<img src="man/figures/README-tassie-3.png" width="100%" />

``` r

prefix_shade <- "WMTS:https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/HillshadeGrey/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"
shade <- wmts(prefix_shade, ortho)
#> zoom: 10
# same: raster::plotRGB(shade, interpolate = TRUE)

prefix_shade <- "WMTS:https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/Hillshade/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"
hillshade <- wmts(prefix_shade, ortho)
#> zoom: 10
raster::plotRGB(hillshade)
```

<img src="man/figures/README-tassie-4.png" width="100%" />

``` r

zlocal <- raster::raster(raster::extent(c(16391003, 16394978, -5301061, -5298768)), crs = raster::projection(ortho))
zz <- wmts(prefix_shade, zlocal)
#> zoom: 14
raster::plotRGB(zz)
```

<img src="man/figures/README-tassie-5.png" width="100%" />

``` r

TasmapRaster <- "https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps/TasmapRaster/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"
xx <- wmts(TasmapRaster, zlocal, max_tiles = 16)
#> zoom: 15
raster::plotRGB(xx)
```

<img src="man/figures/README-tassie-6.png" width="100%" />

``` r


street <- "https://services.thelist.tas.gov.au/arcgis/rest/services/Raster/TTSA/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"
st <- wmts(street, raster::shift(zlocal, 5000, 5000), max_tiles = 36)
#> zoom: 16
raster::plotRGB(st, interpolate = TRUE)
```

<img src="man/figures/README-tassie-7.png" width="100%" />

A list of Tasmanian servers is here:
<https://services.thelist.tas.gov.au/arcgis/rest/services/Basemaps>

## Earthdata

You want the REST GetCapabilities(.xml), and we have to prefix with
“WMTS:” *and include* the layer name
(WIP).

<https://wiki.earthdata.nasa.gov/display/GIBS/GIBS+API+for+Developers#GIBSAPIforDevelopers-ServiceEndpointsandGetCapabilities>

``` r
u <- "WMTS:https://gibs.earthdata.nasa.gov/wmts/epsg3857/best/1.0.0/WMTSCapabilities.xml,layer=MODIS_Terra_L3_Land_Surface_Temp_Monthly_Day"
im <- wmts(u, cbind(0, 0), 20037508, zoom = 0)
#> zoom: 0
library(raster)
#> Loading required package: sp
plotRGB(im, interpolate = TRUE)
```

<img src="man/figures/README-earthdata-1.png" width="100%" />

``` r

im2 <- wmts(u, extent(100, 180, -55, 15), max_tiles= 36)
#> zoom: 4

plotRGB(im2, interpolate = TRUE)
```

## <img src="man/figures/README-earthdata-2.png" width="100%" />

Please note that the wmts project is released with a [Contributor Code
of
Conduct](https://contributor-covenant.org/version/1/0/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md

# wmts 0.0.2

* Original implementation. 


tests/testthat.R
library(testthat)
library(wmts)

test_check("wmts")


tests/testthat/test-wmts.R
u <- file.path("https://basemap.nationalmap.gov/arcgis/rest",
                "services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml")
u1 <- sprintf("WMTS:%s,layer=USGSTopo,tilematrixset=default028mm", u)
test_that("wmts works", {
  rgb <- wmts(u1, cbind(-100, 30)) %>% expect_s4_class("RasterBrick")
  rgba <- wmts(u1, cbind(-100, 30), bands = 1:4, silent = TRUE, max_tiles = 4) %>% expect_s4_class("RasterBrick")
  expect_equal(raster::nlayers(rgba), 4L)
  expect_equal(raster::nlayers(rgb), 3L)

  expect_warning(wmts(u1, cbind(-100, 30)))
  expect_error(wmts(u, cbind(-100, 30), 5000, sds = 3))

 r <- raster::raster(  system.file("extdata/sst.tif", package = "vapour", mustWork = TRUE))
  a <- expect_s4_class(wmts(u1, r, zoom = 3), "RasterBrick")

})


tests/testthat/test-url-handling.R
test_that("URL handling works", {
  skip_on_cran()
  expect_equal(gdalsource_handler("https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml"), 
   "WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer")
  expect_equal(gdalsource_handler("WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer=USGSTopo,tilematrixset=default028mm"), 
               "WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer=USGSTopo,tilematrixset=default028mm")
})


R/spatial_bbox.R
# from ceramic:::is_spatial()
is_spatial <- function (x)
{
  if (inherits(x, "Spatial") || inherits(x, "sf") || inherits(x,
                                                              "sfc") || inherits(x, "BasicRaster") || inherits(x, "Extent")) {
    return(TRUE)
  }
  FALSE
}

# ceramic:::project_spex
project_spex <- function (x, crs)
{
  ex <- c(raster::xmin(x), raster::xmax(x), raster::ymin(x),
          raster::ymax(x))
  idx <- c(1, 1, 2, 2, 1, 3, 4, 4, 3, 3)
  xy <- matrix(ex[idx], ncol = 2L)
  afun <- function(aa) stats::approx(seq_along(aa), aa, n = 180L)$y
  srcproj <- raster::projection(x)
  if (is.na(srcproj)) {
    if (raster::couldBeLonLat(x, warnings = FALSE)) {
      warning("loc CRS is not set, assuming longlat")
      srcproj <- "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0 "
    }
    else {
      stop("loc CRS is not set, and does not seem to be longitude/latitude data")
    }
  }
  raster::extent(reproj::reproj(cbind(afun(xy[, 1L]), afun(xy[,
                                                              2L])), target = crs, source = srcproj)[, 1:2])
}

# ceramic:::spex_to_buff
spex_to_buff <- function (x)
{
  ex <- project_spex(x, "+proj=merc +a=6378137 +b=6378137")
  c(raster::xmax(ex) - raster::xmin(ex), raster::ymax(ex) -
      raster::ymin(ex))
}
# ceramic:::spex_to_pt
spex_to_pt <- function (x)
{
  pt <- cbind(mean(c(raster::xmax(x), raster::xmin(x))), mean(c(raster::ymax(x),
                                                                raster::ymin(x))))
  srcproj <- raster::projection(x)
  if (is.na(srcproj)) {
    if (raster::couldBeLonLat(x, warnings = FALSE)) {
      warning("loc CRS is not set, assuming longlat")
      raster::projection(x) <- "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0 "
    }
  }
  if (!raster::isLonLat(x)) {
    pt <- reproj::reproj(pt, "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0 ",
                         source = raster::projection(x))[, 1:2, drop = FALSE]
  }
  pt
}
#' Spatial bounds
#'
#' Get an extent in longlat ('$tile_bbox') from anything, spatial, raster, sf, extent,
#' point with buffer. Includes user input bounds ('$user_points').
#'
#' Assumes longlat if it seems sensible, from ceramic:::spatial_bbox
#' @noRd
#' @return list with tile_bbox and user_points
spatial_bbox <- function (loc, buffer = NULL)
{
  if (is_spatial(loc)) {
    if (inherits(loc, "Extent")) {
      spx <- spex::spex(loc, crs = "+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs +towgs84=0,0,0")
      if (!raster::couldBeLonLat(spx)) {
        stop("raw extent 'loc' does not seem to be longitude/latitude (use object with CRS)")
      }
    }
    else {
      spx <- spex::spex(loc)
    }
    loc <- spex_to_pt(spx)
    buffer <- spex_to_buff(spx)/2
  }
  if (is.null(buffer))
    buffer <- c(0, 0)
  if (any(!buffer > 0)) {
    if (any(buffer > 0))
      buffer <- rep(max(buffer, 2L))
    if (all(!buffer > 0)) {
      warning("input object has no width or height, using default buffer (5000m)")
      buffer <- c(5000, 5000)
    }
  }
  buffer <- rep(buffer, length.out = 2L)
  if (length(loc) > 2) {
    warning("'loc' should be a length-2 vector 'c(lon, lat)' or matrix 'cbind(lon, lat)'")
  }
  if (is.null(dim(loc))) {
    loc <- matrix(loc[1:2], ncol = 2L)
  }
  loc <- slippymath::lonlat_to_merc(loc)
  xp <- buffer[1]
  yp <- buffer[2]
  bb_points <- matrix(c(loc[1, 1] - xp, loc[1, 2] - yp, loc[1,
                                                            1] + xp, loc[1, 2] + yp), 2, 2, byrow = TRUE)
  if (!slippymath::within_merc_extent(bb_points)) {
    warning("The combination of buffer and location extends beyond the tile grid extent. The buffer will be truncated.")
    bb_points <- slippymath::merc_truncate(bb_points)
  }
  bb_points_lonlat <- slippymath::merc_to_lonlat(bb_points)
  tile_bbox <- c(xmin = bb_points_lonlat[1, 1], ymin = bb_points_lonlat[1,
                                                                        2], xmax = bb_points_lonlat[2, 1], ymax = bb_points_lonlat[2,
                                                                                                                                   2])
  user_points <- bb_points
  list(tile_bbox = tile_bbox, user_points = user_points)
}


R/wmts.R
is_http_generic <- function(x) {
  grepl("^http", x, ignore.case = TRUE)
}
gdalsource_handler <- function(x, sds = 1, ...) {
  ## x might be
  ##      URL https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml
  ##      DSN WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml,layer=USGSTopo,tilematrixset=default028mm
  ## If URL, need SDS choice
  ## (what if it's ecwp:// ?)
  if (is_http_generic(x)) {
    subds <- vapour::vapour_sds_names(x)
    if (sds > 1 || sds < 0 || sds > length(subds[["subdataset"]])) {
      stop(sprintf("'sds' must be 1 or larger, there are %i subdatasets", length(subds[["subdataset"]])))
    }
    if (length(sds) > 1) warning("only one subdataset can be accessed at a time, 'sds[1]' will be used")
    x <- subds[["subdataset"]][sds[1]]
  } else {
    ## we assume you've declared the DRIVER:<url>[,option1][,option2]etc.
  }

  x
}

#' OGC Webmap Tile Server (WMTS)
#'
#' Get raster from WMTS
#'
#' Beware that if 'zoom' is set the 'max_tiles' argument is ignored. Use
#' the function defaults to discover a reasonable zoom level, then re-run with a
#' higher (more pixels) or lower (fewer pixels) zoom level.
#'
#' @param x WMTS address, either the raw 'WMTSCapabilities.xml' or an expanded GDAL 'WMTS:<url>[],options]' driver data source
#' @param loc something we can get an extent from
#' @param buffer radius in metres (in the case 'loc' is a single point, and is longlat)
#' @param silent suppress messages (default is `FALSE`)
#' @param ... ignored
#' @param zoom override hueristic for zoom level (take care! no safety checks)
#' @param max_tiles max number of tiles for zoom hueristic (default 25)
#' @param sds which subdataset to read if there are multiple (defaults to 1)
#' @param bands bands to get (default is 1,2,3 assuming RGB)
#'
#' @export
#' @return raster brick
#' @examples
#' centre <- c(-80.888, 32.332)  ## lonlat
#' radius <- 4000                ## metres
#' u <- file.path("https://basemap.nationalmap.gov/arcgis/rest",
#'                "services/USGSTopo/MapServer/WMTS/1.0.0/WMTSCapabilities.xml")
#' x <- wmts(u, centre, buffer = radius)
#' # u1 <- sprintf("WMTS:%s,layer=USGSTopo,tilematrixset=default028mm", u)
#' # x <- wmts(u1, centre, buffer = radius)
#' # x <- wmts(u, cbind(0, 0), buffer = 20037508)
#' raster::plotRGB(x, interpolate = TRUE) ## use interpolate to match to device size
#' #f <- system.file("gpkg/nc.gpkg", package = "sf", mustWork = TRUE)
#' #sf <- sf::read_sf(f)
#' #x <- wmts(u, sf)
#'
#'
#' #tab <- tibble::as_tibble(raster::as.data.frame(x, xy = TRUE))
#' #names(tab) <- c("x", "y", "red", "green", "blue")
#' ##tab <- dplyr::filter(tab, !is.na(red))
#' #
#' ### ... when we have missing values, we should drop them or rgb() will error
#' #tab$hex <- rgb(tab$red, tab$green, tab$blue, maxColorValue = 255)
#' #library(ggplot2)
#' #ggplot(tab, aes(x, y, fill = hex)) +
#' #   geom_raster(interpolate = TRUE) +
#' #   coord_equal() +
#' #   scale_fill_identity()
wmts <- function(x, loc, buffer = NULL, silent = FALSE, ..., zoom = NULL, max_tiles = 9, bands = 1:3, sds = 1L) {
  x <- gdalsource_handler(x, sds = sds)
  if (is.numeric(loc)) loc <- matrix(loc, byrow = TRUE, ncol = 2)
  bbox_pair <- spatial_bbox(loc, buffer)
  my_bbox <- bbox_pair$tile_bbox
  bb_points <- bbox_pair$user_points
  ## all we need is zoom
  if (!is.null(zoom))   max_tiles <- NULL
  tile_grid <- slippymath::bbox_to_tile_grid(my_bbox, max_tiles = max_tiles,
                                           zoom = zoom)
  zoom_level <- tile_grid$zoom
  if (!is.null(zoom)) {
    zoom_level <- zoom
  }
  if (!silent) {
    writeLines(sprintf("zoom: %s", zoom_level))
  }
  uzoom <- paste0(x, sprintf(",zoom_level=%i", zoom_level))
  ex <- raster::extent(slippymath::lonlat_to_merc(as.matrix(expand.grid(x = my_bbox[c(1, 3)], y = my_bbox[c(2, 4)]))))
 br <- raster::subset(raster::brick(uzoom), bands)
    out <- try(raster::crop(br, ex, snap = "out"), silent = TRUE)  ## FIXME how to avoid raster creating a file?

  if (inherits(out, "try-error")) {
    stop(sprintf("cannot read from WMTS url: %s", uzoom))  ## FIXME add longlat-bounds to message
  }
  suppressWarnings(raster::readAll(out))
}


